src/object/cli_ec.c

```diff
@@ -33,6 +33,1086 @@
+		if (recx_end > full_end) {
+			tmp_recx.rx_idx = full_end;
+			tmp_recx.rx_nr = recx_end - full_end;
+			ec_data_recx_add(&tmp_recx, riod->iod_recxs, ridx,
+					 tgt_recx_idxs, oca, true);
+			ec_data_seg_add(&tmp_recx, iod_size, sgl, &iov_idx,
+					&iov_off, oca, iovs, iov_nr, sorter,
+					true);
+		}
+	}
+
+	for (i = 0; i < ec_recx_array->oer_nr; i++) {
+		full_ec_recx = &ec_recx_array->oer_recxs[i];
+		full_recx = &full_ec_recx->oer_recx;
+		ec_parity_recx_add(full_recx, riod->iod_recxs, ridx,
+				   tgt_recx_idxs, oca);
+	}
+	ec_parity_seg_add(ec_recx_array, iod, oca, sorter);
+
+	obj_ec_seg_pack(sorter, rsgl);
+
+	/* generate the oiod/siod */
+	for (i = 0, idx = 0, last = 0; i < obj_ec_tgt_nr(oca); i++) {
+		/* get each tgt's idx in the compact oiod_siods array */
+		if (tgt_recx_nrs[i] != 0)
+			tidx[i] = idx++;
+		else
+			tidx[i] = -1;
+		for (j = last; j < tgt_recx_idxs[i] + tgt_recx_nrs[i]; j++) {
+			if (riod->iod_recxs[j].rx_nr != 0)
+				continue;
+			/* being merged so left empty space */
+			D_ASSERT(j != tgt_recx_idxs[i]);
+			D_ASSERT(j < riod->iod_nr);
+			for (k = j; k < tgt_recx_nrs[i] + tgt_recx_idxs[i]; k++)
+				D_ASSERT(riod->iod_recxs[k].rx_nr == 0);
+			empty_nr = tgt_recx_nrs[i] + tgt_recx_idxs[i] - j;
+			for (k = j; k < riod->iod_nr - empty_nr; k++)
+				riod->iod_recxs[k] =
+					riod->iod_recxs[k + empty_nr];
+			for (k = riod->iod_nr - empty_nr;
+			     k < riod->iod_nr; k++) {
+				riod->iod_recxs[k].rx_idx = 0;
+				riod->iod_recxs[k].rx_nr = 0;
+			}
+			tgt_recx_nrs[i] -= empty_nr;
+			for (k = i + 1; k < obj_ec_tgt_nr(oca); k++)
+				tgt_recx_idxs[k] -= empty_nr;
+			riod->iod_nr -= empty_nr;
+			break;
+		}
+		last = tgt_recx_idxs[i] + tgt_recx_nrs[i];
+	}
+	oiod->oiod_nr = idx;
+	for (i = 0, rec_nr = 0, last = 0; i < obj_ec_tgt_nr(oca); i++) {
+		if (tgt_recx_nrs[i] == 0)
+			continue;
+		siod = &oiod->oiod_siods[tidx[i]];
+		siod->siod_tgt_idx = i;
+		siod->siod_idx = tgt_recx_idxs[i];
+		siod->siod_nr = tgt_recx_nrs[i];
+		siod->siod_off = rec_nr * iod_size;
+		for (idx = last; idx < tgt_recx_idxs[i] + tgt_recx_nrs[i];
+		     idx++) {
+			rec_nr += riod->iod_recxs[idx].rx_nr;
+		}
+		last = tgt_recx_idxs[i] + tgt_recx_nrs[i];
+	}
+
+#if EC_DEBUG
+	obj_reasb_req_dump(reasb_req, sgl, oca, stripe_rec_nr, iod_idx);
+#endif
+
+	if (iovs != NULL && iovs != iov_inline)
+		D_FREE(iovs);
+	return rc;
+}
+
+int
+obj_ec_req_reassemb(daos_obj_rw_t *args, daos_obj_id_t oid,
+		    struct daos_oclass_attr *oca,
+		    struct obj_reasb_req *reasb_req)
+{
+	daos_iod_t		*iods;
+	d_sg_list_t		*sgls;
+	uint32_t		 iod_nr = args->nr;
+	int			 i, rc = 0;
+
+	iods = args->iods;
+	sgls = args->sgls;
+	for (i = 0; i < iod_nr; i++) {
+		rc = obj_ec_recx_scan(&iods[i], &sgls[i], oca, reasb_req, i);
+		if (rc) {
+			D_ERROR(DF_OID" obj_ec_recx_scan failed %d.\n",
+				DP_OID(oid), rc);
+			goto out;
+		}
+
+		rc = obj_ec_recx_encode(oid, &iods[i], &sgls[i], oca,
+					&reasb_req->orr_recxs[i]);
```

Comment:
Hmm, it looks like to me  it should encode after scan all of iods[i]? since the stripes might in different iod. or I miss sth?
TBH, I still thought split first, then do encoding will make thing clearer. Still trying to understand the benefits of this obj_ec_recx. Might miss sth.

Reply:
Hmm, I think split iod/sgl first, then sort/encoding can save a lot efforts to resolve those corner cases.  For example
1, Multiple same akeys in the same iod, i.e. you do not need merge them.
2. Get rid of ec_recxs, i.e. you do not need track full stripe during the split.
3. It can decide whether to encode or replication at the encode phase, which give us a lot flexibility.

Reply:
> Hmm, I think split iod/sgl first, then sort/encoding can save a lot efforts to resolve those corner cases. For example

Right now I really don't think split first is better. But things can be clearer during time passing.

> 1, Multiple same akeys in the same iod, i.e. you do not need merge them.

Sorry not sure what is means, there is only one akey per iod.

> 2. Get rid of ec_recxs, i.e. you do not need track full stripe during the split.

It actually make the handling faster by recognizing full stripe first, we may need not pay much attention for this now.

> 3. It can decide whether to encode or replication at the encode phase, which give us a lot flexibility.

I agreed that if later we decided to dispatch at client-side for dtx update, we can delay the encode. But it can be done also on current code.

Reply:
> Right now I really don't think split first is better. But things can be clearer during time passing.

If we split all of recxs/sgls,  a lot corner cases can be resolved, for example if there multiple same akeys in one input.

> 
> > 1, Multiple same akeys in the same iod, i.e. you do not need merge them.
> 
> Sorry not sure what is means, there is only one akey per iod.

Sorry, I mean input here.

> 
> > 1. Get rid of ec_recxs, i.e. you do not need track full stripe during the split.
> 
> It actually make the handling faster by recognizing full stripe first, we may need not pay much attention for this now.

hmm, I think just walking through the split recxs to know the full stripe status is simpler and clearer.   And also this "recognizing" require extra efforts to re-order and merging the recxs, and also adding extra data structures.

> 
> I agreed that if later we decided to dispatch at client-side for dtx update, we can delay the encode. But it can be done also on current code.

Reply:
> If we split all of recxs/sgls, a lot corner cases can be resolved, for example if there multiple same akeys in one input.

Ever discussed this with Liang, we probably can treat it as input error that with same Akey for different iods, at least need not pay much attention for that cornel case for now.

> hmm, I think just walking through the split recxs to know the full stripe status is simpler and clearer. And also this "recognizing" require extra efforts to re-order and merging the recxs, and also adding extra data structures.

TBH, I don't think that is better, I don't know why you prefer a smaller granularity sorting and for a cornel case. But probably need not argue for it now, we can refine it later when needed if it become clearer that it will be better.

Reply:
> Ever discussed this with Liang, we probably can treat it as input error that with same Akey for different iods, at least need not pay much attention for that cornel case for now.

But we still have to check? I mean we should avoid these corner cases in any cases.

> 
> > hmm, I think just walking through the split recxs to know the full stripe status is simpler and clearer. And also this "recognizing" require extra efforts to re-order and merging the recxs, and also adding extra data structures.
> 
> TBH, I don't think that is better, I don't know why you prefer a smaller granularity sorting and for a cornel case. But probably need not argue for it now, we can refine it later when needed if it become clearer that it will be better.

hmm, my major argument is the complexity and constrain of the implementation
 1. Unnecessarily tracking the full stripe status.
 2. required checking and ordering iods before splitting.

If the only benefit of this implementation is just providing larger granularity of sorting,  I do not think it is worth to adding/enforcing these.  Plus there are other potential benefits if we split the data at the beginning for example sharing some of code with array, forward RPC from clients etc.

Reply:
> hmm, my major argument is the complexity and constrain of the implementation
> 
> 1. Unnecessarily tracking the full stripe status.
> 2. required checking and ordering iods before splitting.
> 
> If the only benefit of this implementation is just providing larger granularity of sorting, I do not think it is worth to adding/enforcing these. Plus there are other potential benefits if we split the data at the beginning for example sharing some of code with array, forward RPC from clients etc.

TBH, actually I think it make things much simpler and more efficient by using a very simple data structure and scan first. 
For "sharing some of code with array" it is an unclear requirement till now, I think need not make things complex based on a maybe or may not  requirement.
"forward RPC from clients" it can be easily refined based on current code if that become true later.

Reply:
Sorry, I do not understand why this is simpler and efficient, it needs checking/merging/sorting those iods first, then start scanning to know full stripe status ...

To me, these are unnecessary,  I do not see much benefits to do this.  The full stripe status can be tracked/checked by per target sgl/iod naturally, when we calculate the EC, plus those potential benefits.

Anyway we do not have to decide now, but I would really suggest to avoid those pre-iods handling, whether or not they are rare cases, but it add extra complexity, which is really unnecessary.